File: monorepo_kotlin/apps/web/src/main/kotlin/mono/web/Scopes.kt
// Labeled scope test cases for Kotlin — monorepo edition.
// Cross-package refs target mono.shared.SharedStatus + mono.shared.scopes.ns.Widget.

package mono.web

import mono.shared.SharedStatus as CrossStatus      // S09.def
import mono.shared.SharedStatus
import mono.shared.scopes.ns.Widget                  // S14.import

var moduleVar: String = "mod-initial"                // S04.def

fun s01Local() {
    val localA = "S01.local"                         // S01.def
    println(localA)                                  // S01.read
}

fun s02ClosureRead() {
    val outerA = "S02.outer"                         // S02.outer.def
    val inner = { println(outerA) }                  // S02.inner.read
    inner()
}

fun s03ClosureWrite(): Int {
    var counter = 0                                  // S03.outer.def
    val bump = { counter += 1 }                      // S03.inner.write
    bump(); bump()
    return counter                                   // S03.outer.read
}

fun s05SameModuleWrite() {
    moduleVar = "rotated"                            // S05.write
    println(moduleVar)                               // S05.read
}

fun s06CrossRead(): String = SharedStatus            // S06.read

fun s07CrossWrite() {
    SharedStatus = "S07"                             // S07.write
}

fun s08Shadowing() {
    val moduleVar = "shadowed"                       // S08.shadow.def
    println(moduleVar)                               // S08.shadow.read
}

fun s09AliasedImport() {
    println(CrossStatus)                             // S09.read
}

open class ScopeBase(val x: Int) {                   // S11.instance.def (x)
    companion object { var staticX: Int = 1 }        // S12.static.def / S13.base.def

    fun readInstance(x: Int): Int =
        x + this.x                                   // S11.param.read + S11.instance.read
}

class ScopeDerived : ScopeBase(5) {
    fun readInherited(): Int = staticX               // S13.derived.read
}

fun s14Qualified(): String = Widget("hi").label      // S14.read

fun runScopeDemo() {
    s01Local()
    s02ClosureRead()
    println(s03ClosureWrite())
    s05SameModuleWrite()
    println(s06CrossRead())
    s07CrossWrite()
    s08Shadowing()
    s09AliasedImport()
    println(ScopeBase(42).readInstance(100))
    println(ScopeDerived().readInherited())
    println(s14Qualified())
}

File: monorepo_typescript/apps/web/src/scopes.ts
// Labeled scope test cases for TypeScript — monorepo edition.
// Ported from typescript/scopes.ts. Cross-module refs now target the
// @mono/shared and @mono/utils workspace packages via path aliases.
// See SCOPE_TEST_SPEC.md at repo root.

import { formatUser as sharedFormatUser } from "@mono/shared";     // S09.def
import * as shared from "@mono/shared";                             // for S06/S07
import { reExportedValue } from "./scopes_reexport";                // S10.consumer.import
import { Widget } from "./scopes_ns/inner";                         // S14.import

// ------------------------------------------------ S04 def / S05 write target
export let moduleVar = "mod-initial";                               // S04.def

export function s01Local(): void {
    const localA = "S01.local";                                     // S01.def
    console.log(localA);                                            // S01.read
}

export function s02ClosureRead(): void {
    const outerA = "S02.outer";                                     // S02.outer.def
    const inner = () => console.log(outerA);                        // S02.inner.read
    inner();
}

export function s03ClosureWrite(): number {
    let counter = 0;                                                // S03.outer.def
    const bump = () => { counter++; };                              // S03.inner.write
    bump(); bump();
    return counter;                                                 // S03.outer.read
}

export function s05SameModuleWrite(): void {
    moduleVar = "rotated";                                          // S05.write
    console.log(moduleVar);                                         // S05.read
}

export function s06CrossRead(): string {
    return shared.DEFAULT_ROLE;                                     // S06.read
}

export function s07CrossWrite(): void {
    // Writing to an exported const isn't legal TS, so write to a
    // mutable namespace member instead. Exercises cross-pkg write.
    (shared as unknown as { DEFAULT_ROLE: string }).DEFAULT_ROLE = "S07"; // S07.write
}

export function s08Shadowing(): void {
    const moduleVar = "shadowed";                                   // S08.shadow.def
    console.log(moduleVar);                                         // S08.shadow.read
}

export function s09AliasedImport(): void {
    console.log(sharedFormatUser({ id: 1, name: "s09" }));          // S09.read
}

export function s10ReexportChain(): void {
    console.log(reExportedValue);                                   // S10.consumer.read
}

export class ScopeBase {
    static staticX = 1;                                             // S12.static.def / S13.base.def
    constructor(public x: number) {}                                // S11.instance.def (x)

    readInstance(x: number): number {
        return x + this.x;                                          // S11.param.read + S11.instance.read
    }
}

export class ScopeDerived extends ScopeBase {
    constructor() { super(5); }
    readInherited(): number {
        return ScopeBase.staticX;                                   // S13.derived.read
    }
}

export function s14Qualified(): string {
    return new Widget("hi").label;                                  // S14.read
}

export function runScopeDemo(): void {
    s01Local();
    s02ClosureRead();
    console.log(s03ClosureWrite());
    s05SameModuleWrite();
    console.log(s06CrossRead());
    s07CrossWrite();
    s08Shadowing();
    s09AliasedImport();
    s10ReexportChain();
    console.log(new ScopeBase(42).readInstance(100));
    console.log(new ScopeDerived().readInherited());
    console.log(s14Qualified());
}
